### ISSUE TITLE
from on association ignored when eager loading

### ISSUE THREAD

**Issue by jturkel** [rails/rails#15985]
Any `from` specified during on association scopes is ignored during eager loading. Consider the following model:

```
class Blog < ActiveRecord::Base
  has_many :posts, -> { from('(select posts.* from posts where posts.title is not null) posts') }
end
```

Loading this association without eager loading runs the following query:

```
SELECT "posts".* FROM (select posts.* from posts where posts.title is not null) posts WHERE "posts"."blog_id" = ?  [["blog_id", 1]]
```

Eager loading the association results in the following query:

```
SELECT "posts".* FROM "posts" WHERE "posts"."blog_id" IN (1)
```

Full test case can be found here: https://gist.github.com/jturkel/e3850efcd8d699f99b85

**Comment by mockdeep** [comment]
We've run into something similar in our codebase. Not sure if it's the exact same thing, but seems related:

```ruby
has_many :early_responses,
  -> { where(forms: { version: 1 }).order(:created_at) },
  through: :forms,
  source: :responses
```

If I remove the `source:` and tack on a `joins(:forms)`, it seems to work as expected.

**Comment by mockdeep** [comment]
If I distribute the relationship across 2 models, it seems to resolve correctly:

```ruby
class User
  has_many :early_responses,
    -> { where(forms: { version: 1 }).order(:created_at) },
    through: :forms
end

class Form
  has_many :early_responses, class_name: 'Response'
end
```

**Comment by Grammarella** [comment]
We've seen something similar, too.  This is the association:

`has_many :successor_tags, ->{order(tag_text: :asc)}, through: :successor_connections, source: :successor`

Doing preloading (with includes) loses the default ordering--it doesn't even use that order clause in the lambda when the query is fired.  Reloading the tag or rescoping it as suggested it does fix the problem, but by hitting the database again to do a fresh load.  It seems that this happens because we are using the source param in the has_many.  So there are workarounds, but the behavior still seems unexpected.

**Comment by devonestes** [comment]
I'm seeing this same behavior in 5.0.2. If no one is working on this, then I might take a swing at it.

**Comment by kamipo** [comment]
This should be fixed by #29413.
